### DISCUSSION TITLE
Preferred way to update dependency

### DISCUSSION THREAD

**Discussion by mdbergmann** [ocicl/ocicl#49] [General]
Hello.

I'd like to know to would I update a local dependency of a project.

I see a few possible ways:

1. use version specifier in .asd file. Upon `asdf:load-system` it seems that this doesn't trigger re-downloading and re-populating systems.csv
2. modify systems.csv and call `ocicl install`.
3. use `ocicl install system:version` to update a system

Not sure if 2) actually works, haven't tried but I suppose it would because it should be possible to install/download all dependencies from just systems.csv.

**Comment by mdbergmann** [comment]
I'm asking because I know from Erlang/Elixir, or Java (Maven, Gradle), etc. that there is exactly one place where dependencies are defined which is the main project config/description file. Dependencies are defined there with a version number (or just latest if specific version number is needed). Then the tooling will make sure that dependencies are installed/updated (or older ones removed).

In CL this file for me is the .asd file. But yet, when specifying system dependencies in there the version is ignored by ocicl (or is it?) and just latest versions are installed by ocicl if none has been installed prior via ocicl command-line tool.

So I take it that dependencies should be managed (installed/updated/removed) via the ocicl command-line tool. But that makes two places that need to be touched when changing dependencies.

**Comment by mdbergmann** [comment]
There is another 'thing' I'd need to ask about.

Often, when developing, you want to load a library and try it out whether it is the right one. Maybe there are alternatives, so you end up loading a few. Assuming quicklisp is disabled (or not installed), you would do that with ASDF:LOAD-SYSTEM which implicitly loads the library via ocicl. But ocicl will also add the library to systems.csv, which at this point it should not, only when I really declare the library as a system dependency in the .asd.

**Comment by atgreen** [comment]
You should never have to do [2].   [3] is the way to do things.  Maybe [1] could be made to work someday.
I wouldn't think of systems.csv as where dependencies are defined.  It only specifies what's been downloaded and available for use.  Project dependencies are still specified in your .asd files.  You should be able to delete systems.csv at any time.  If you have extra items in there.. it shouldn't matter at all, unless you are also distributing your systems.csv file, in which case I suppose you may want to prune it.    If you have some suggestions on how to improve this, please let m know.  I didn't really think much about specifying versions in the .asd files, because I don't see that very often in the wild.

**Reply by mdbergmann** [reply]
OK, thanks.
I was indeed thinking that systems.csv should mirror very closely the dependencies of the project, and hence I thought of distributing systems.csv.

But probably I looked at ocicl from wrong angle. If looking at it as a loose _local_ systems repo then it's fine.
